### ISSUE TITLE
Overall Testing Story for Office Add-ins

### ISSUE THREAD

**Issue by LanceEa** [OfficeDev/office-js-docs-pr#584]
*This potentially is a User Voice but thought I would start here*

**TL;DR**

- __Challenge 1__: No documentation, or direction on mocking Office-Js Api's for testing.
I know there are many permutations and opinions on tooling and testing frameworks so documenting them all is not feasible. But, potentially providing a mock or at least a small write up showing some samples I think would help.

- __Challenge 2__: No easy way to test against multiple platforms/versions of Excel
 It would be nice if there was an easier way to test against multiple platforms. Similar to BrowserStack, SauceLabs and Xamarin Test Cloud or at something like what the Edge Team did with Virtual Machines. This is probably more of a User Voice but also could benefit from a short write-up giving some guidance.

**Issue:** 
There is no documentation on how to best test add-in code. I understand that is fairly broad so let me explain.

I'm developing my add-in with Angular which has a great Unit and E2E integration testing story. Since the add-ins is web based I can leverage the Angular teams opinions on testing and if someone was using some other framework they could follow the guidelines set by that framework. But, where it gets a little bit dicey is when you start interacting with Office-js API's. 

So, what I did was first focus completely on the UI and aspects of add-in that do not require office-js. This way I can follow my normal developer workflow (Chrome Debug tools, HMR code refreshes, automated unit testing,etc...).  Then I introduced the Office API's and at this point I have avoided writing any tests for those functions. To me that is the first challenge because there is no examples, no mocks and in general no direction on how to best test these. I recognize I could mock/spy on the office api's but the setup to do that would be fairly painful (I recognize someone has to feel the pain at least once even if a mock was provided 😄 ).

The other challenge is when we need to actually test the add-in within Excel. I have to physically test on my Mac, my Windows 10 PC and with different version of Excel. The number of combinations of OS/Excel version is definitely not controllable, especially with external users (outside our company). So, I would love to get some direction here.

**Comment by sancarn** [comment]
[Open office has javascript capability](https://www.openoffice.org/framework/scripting/release-0.2/javascript-devguide.html) Might be possible to build a OfficeJS wrapper for open office. That'd provide a pretty good testing ground for developers at least.

**Comment by mattcopas** [comment]
+1 - some guidelines/examples on unit testing would be much appreciated :)

**Comment by shirbr510** [comment]
I tried mocking Office with some sort of an `office-js` wrapper I've written but so far have failed in doing so (had my whole mocks working but my app couldn't work with Officejs properly)

Would also love to see some sort of example or documentation

**Comment by shirbr510** [comment]
I have ended up doing the following @  `setupTests.js` (I'm using jest if it matters):

    import OfficeMock from './__mocks__/Office'
    
    const Office = global.Office;
    
    beforeEach(()=>{
        global.Office = OfficeMock
    })
    
    afterEach(()=>{
        global.Office = Office
    })

the `OfficeMock` is a simple object mapping that fits the properties I need from Office.js

I do believe that a proper solution including some guidelines and examples is still necessary though...

**Comment by AlekseyPi** [comment]
Test:

`/* eslint-disable @typescript-eslint/no-angle-bracket-type-assertion */
/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
import ParagraphExtensions from "../office_extensions/ParagraphExtensions";
import {TextRange} from "../models/TextRange";

test("getCustomRange", async() => {
  const expectedRange = <Word.Range>{text: "  "};

  const paragraph: Word.Paragraph = <Word.Paragraph>{
    text: "text  text text  text text",
    search: function(_text) {
      return <Word.RangeCollection>{
        load: function() {
          return this;
        },
        items: [null, expectedRange]
      };
    },
    context: {
      sync: function() {
        return new Promise(accept => accept());
      }
    }
  };

  const textRange: TextRange = {
    start: 15,
    length: 2
  };

  expect(await ParagraphExtensions.getCustomRange(paragraph, textRange)).toBe(
    expectedRange
  );
});
`

Class:

`import {TextRange} from "../models/TextRange";

export default class ParagraphExtensions {
  public static async getCustomRange(
    paragraph: Word.Paragraph,
    textRange: TextRange
  ): Promise<Word.Range> {
    const text = paragraph.text.substring(
      textRange.start,
      textRange.start + textRange.length
    );

    let textIndex = 0;
    let i = 0;

    while ((textIndex = paragraph.text.indexOf(text, textIndex)) != -1) {
      if (textIndex == textRange.start) {
        break;
      }
      i++;
      textIndex += textRange.length;
    }

    const ranges = paragraph.search(text).load("items");
    await paragraph.context.sync();

    return ranges.items[i];
  }
}`

**Comment by AlekseyPi** [comment]
```

**Comment by Alex-ray** [comment]
Hello @kbrandl , I also need support for this. 

My team is building a word add-in that helps lawyers create templates for their documents. Naturally reliability is our number one priority here. End to end integration tests are critical in our ability to ensure our add-in's reliability. 

The actual work this might take from your team feels low in relation to the potential increase in quality and reliability of your add-ins.

**Comment by kbrandl** [comment]
Thanks for reaching out @Alex-ray and I can certainly appreciate the perspective you've shared. This work item is in the documentation backlog; @davidchesnut is there any update you can provide regarding status of this work item?

**Comment by davidchesnut** [comment]
Hi @kbrandl @LanceEa and folks,
The engineering team is looking into solutions in this space, but of course it needs to be general and scalable for multiple testing frameworks. So that effort will take some time. On the documentation side I believe there are some patterns we can share to make it easier to test. I'm working to get those documented and will provide an update once we have something available. Thanks!

**Comment by phlppnhllngr** [comment]
If anyone is interested, I created a small demo/repository where I wrote down some instructions on how to implement e2e tests for TaskPane apps (in this case the sample Contoso app for Word created by yeoman generator office). I was using Appium and WinAppDriver. 
The link is: https://github.com/phlppnhllngr/office-addin-integration-testing

**Comment by kb339** [comment]
How to integrate Office.js with UFT/Selenium for Excel online (Office 365) automation testing?

**Comment by seifeldeen92** [comment]
@kb339 Have you reached any progress in that? i am stuck in the same spot

**Comment by RohilaDasari** [comment]
Is there any update on this?

**Comment by JensMadsen** [comment]
Any updates? We just deployed our add-in with a huge bug in it ... We need automated tests

**Comment by celownick** [comment]
Hi @davidchesnut , @kbrandl and folks, 
can I kindly ask for any updates on this ? We are preparing strategy for add-in testing and answering questions mentioned by @LanceEa wound have a crucial meaning for our team. 
Thank you

**Comment by JensMadsen** [comment]
Any updates? We need to know the timeline for this "feature". If you know that this will happen in 2025 let us know. In that case, we will invest time in developing e.g. Selenium click e2e tests. If you, on the other hand, are actually working on this an expect a release in 2020 we will probably wait for it :-) But without an update, it is difficult to make a strategic decision on what to do (we must have test for CI/CD and QA). Best Regards

**Comment by ludmilanesvitiy** [comment]
Hey team! We also waiting for some updates with examples for e2e testing of word add-in. Please, post some updates - when do you plan to create this documentation? Is it possible at all to make this type of automated testing to work with word add-in. If I can help you with something relative to it to speed up the issue resolving - just let me know.

**Comment by ElizabethSamuel-MSFT** [comment]
Related issue: https://github.com/OfficeDev/office-js-docs-pr/issues/1611

**Comment by JensMadsen** [comment]
Any news? @ElizabethSamuel-MSFT

**Comment by JensMadsen** [comment]
We have started building ui tests using Selenium. We currently only test Word online. We plan to use Winium to test Word desktop Windows. The tests run on Github Actions and run headless in incognito mode. Currently, we only have a setup for Chrome (chromedriver). If you plan to create a similar setup please note that: 

1. In order for the addin to load and function properly you have to change the default Selenium cookie settings when running in incognito mode: 
``` typescript
chromeOptions.setUserPreferences({ 'profile.cookie_controls_mode': 0 })
``` 
We had some problems before we realized this because Word Online loads fine and you will be able to login, but the addin does not show up anywhere (ribbon, appsource, edit->my addins, etc). 

2. For stable tests you have to fix the window sizes by running: 
``` typescript
await driver
      .manage()
      .window()
      .setSize(3000, 2133)
```
(`setRect` for newer versions of Selenium).

3. The addin runs in a designated iframe which is nested in the application iframe. In order to interact with the addin you have to switch to the parent iframe and then locate and switch to the addin iframe: 
``` typescript
// switch to top iframe
await driver.wait(until.ableToSwitchToFrame(0))
// locate addin iframe
const addinIframes = await driver.findElements(
      By.xpath('//iframe[contains(@title,"<name of your addin>")]'),
    )
await driver.wait(until.ableToSwitchToFrame(addinIframes[0]))
```

Hope this can be of some help. If you need more details please ask :-)

**Comment by piemart** [comment]
Hi @JensMadsen, 
Thanks for the info. 
I would like to setup UI testing with Selenium on an Outlook Desktop add-in with edge chromium, but I'm struggling. 
The add-in starts the localhost:3000/taskpane.html and after some user inputs, the taskpane window opens an external appdomain (listed in the manifext.xml file). Here I would like Selenium to take over, so I can select the desired elements by their css_selector. 
Do you think this is possible? Some more details would be appreciated :-)

**Comment by JensMadsen** [comment]
Hi @piemart 

No quite sure what you wnat to achieve, but it sounds similar to what I do. In metasteps:
1. start in incognito mode (I use Chrome). Incognito more stable
2. login to office.com (make sure you have a test user). you have to pass username and password
3. I open an existing test document (I test Word)-> office  opens in a new tab
4. go to new tab
5. Do your tests (I test a taskpane addin; the one that load on the side )

**Comment by abdulkareemnalband** [comment]
We would like to use https://github.com/microsoft/playwright to automate the testing of desktop addins.
If word addin will expose WebView2 instance over Chrome DevTools Protocol , then we could use 
https://playwright.dev/docs/api/class-browsertype#browser-type-connect-over-cdp to attach to that and use playwright to write tests

**Comment by mitoihs** [comment]
My coworker has recently set up Selenium for desktop Excel add-in. That's what he said:

> Since a lot of people are struggling, I guess I can give some information. Please note that this is a whacky way of doing it, but AFAIK only way for now.
> 
> Webview2 (chromium edge) has support for Selenium: https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver
> It is possible to instrumentate desktop Office (in my case Excel) to start Webview2 and connect with Selenium. To instrumentate Excel, I've used WinAppDriver. It is hard to set up (notable problems are WinAppDriver being slow and not suitable to drive Excel over simple clicks, WinAppDriver requiring Appium requiring Selenium v3, but Webview2 needs Selenium v4 - I've overcome this by using multiple projects that only interface externally).
> If all goes correctly you will end up with WinAppDriver application that can start Office and connect Webview2, and Selenium that can drive add-in. It is hard to control the actual application, but for that purpose, I use OfficeJs via Selenium function ExecuteScript.
> 
> TL;DR: Possible to use Desktop Office with Selenium, very hard to set up.

**Comment by ferrykranenburgcw** [comment]
We would like to test our add-in in Outlook web using Playwright. But it looks very hard to just activate an e-mail and then activate our add-in inside Outlook. We also tried another way to test our code in Javascript only (using the Jest test framework) but we are missing bits an pieces from the Office.js framework to make it all work.

Any documentation about unit testing and integration testing add-in would be very welcome and helpful.

**Comment by AlexJerabek** [comment]
Hi everyone,

We appreciate all the discussion and feedback requesting testing solutions for Office Add-ins and Office JS. Unfortunately, there are no automated or unit testing solutions that we provide at this time. We're working on a solution, but have no concrete implementation or release details to share at this time. We realize this is an unsatisfying answer, but we feel it’s better to be clear about the support story, rather than to create false expectations.

We are moving the discussion to M365 Developer Platform Ideas. For the primary concern around standardized testing, please upvote and comment on this idea: [Standard way to unit test and mock office-js API](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/standard-way-to-unit-test-and-mock-office-js-api/idi-p/2718673). For the other related requests that have accumulated on this issue, we also ask that you take your requests to the [M365 Developer Platform Ideas](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/idb-p/Microsoft365DeveloperPlatform/label-name/Office%20Add-ins%20development) and submit them as ideas (or upvote existing ones). This will help us prioritize things in an orderly fashion.
 
If there are other issues about the documentation that have been raised in this thread but were not addressed, please create a new issue to track each specific problem. 
 
Thank you.

**Comment by lafe** [comment]
@AlexJerabek I think, this is a very unsatisfying approach on how to deal with this issue. It has 42 comments and has been open for almost four years. By creating a new issue on a different platform, the complete history is gone (the new idea does not even reference this GitHub issue) and it appears as a new request. However, it is not a new customer request but an RFC from Microsoft to the community. Instead of posting it here and engaging directly with the affected developers, it is moved to a different platform and poses some questions. To me it appears, that Microsoft does not want to deal with this issue and moving it to a different platform buys time. Furthermore, if there is no feedback, you can internally point to the new idea and say "we tried but there seems to be no interest".

Furthermore, the new platform seems to be one where ideas go to die: currently, there are 217 ideas with the state "New" but only 4 ideas that have a different state (2 in "Working on it" and 2 in "Completed"). The ideas with different state where updated by their creators, so those are not community requests that were introduced through this platform. Even off-topic ideas like [this one](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/xbox-one-games/idi-p/2361010#M55) are not moderated. An actively used platform that is used for product planning and is actively monitored looks different...

I can somewhat understand, if you do not want to deal with this issue, as it is complex and might only be relevant for a subset of users. However, please be honest with it instead of pulling moves like this.

**Comment by giulianozor** [comment]
@lafe I agree with you. IMHO the project is (almost) dead and I see very little chances of it being maintained unless some major security issue is found. Unfortunately there are no alternatives AFAIK

**Comment by Rick-Kirkham** [comment]
@lafe @giulianozor We now have documentation of unit testing and mocking the Office.js.

[Unit testing in Office Add-ins](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/unit-testing)

**Comment by rootlocus22** [comment]
Is there a Cypress like solution for Office Add In, we need to prepare an end to end integration testing framework to automate our flows?

**Comment by igor-ribeiiro** [comment]
> Is there a Cypress like solution for Office Add In, we need to prepare an end to end integration testing framework to automate our flows?

Unfortunately, we currently don't have any end-to-end testing framework. So we don't have a Cypress-like solution at the current time.

**Comment by Rinzwind** [comment]
As a few others have commented here on using Selenium WebDriver for UI testing add-ins: one problem we’re bumping into is that Excel on office.com pops up a dialog saying “Looking to share a section of your workbook?” (see screenshot below) which covers the top part of our add-in’s task pane, so when a test tries to click an element there it gets an [‘element click intercepted’](https://www.w3.org/TR/2018/REC-webdriver1-20180605/#dfn-element-click-intercepted) error. As the dialog doesn’t seem to pop up consistently at the same point in the tests, I’m not quite sure yet how to deal with it. I’m wondering whether others have bumped into this and can offer advice? More generally: is there a more suitable place than this issue to ask questions about UI testing add-ins?

![Dialog Share Section of Workbook](https://user-images.githubusercontent.com/1611248/228931745-43f23039-9085-4ec2-a9ff-b28e81b2ef77.png)

**Comment by AlexJerabek** [comment]
> is there a more suitable place than this issue to ask questions about UI testing add-ins?
@Rinzwind, Stack Overflow is the best way to get answers from others in the same situation. Stack's great for the intersection of technologies, as opposed to pure Office.js questions.
